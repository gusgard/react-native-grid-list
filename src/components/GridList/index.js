import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Animated, ViewPropTypes, View, FlatList } from 'react-native';

import { colors } from '../../themes';
import generateStyles from './styles';

export default class GridList extends PureComponent {
  static propTypes = {
    animationDuration: PropTypes.number,
    animationInitialBackgroundColor: PropTypes.string,
    itemStyle: ViewPropTypes.style,
    numColumns: PropTypes.number.isRequired,
    separatorBorderColor: PropTypes.string,
    separatorBorderWidth: PropTypes.number,
    showAnimation: PropTypes.bool,
    showSeparator: PropTypes.bool,

    data: PropTypes.array,
    renderItem: PropTypes.func,

    // Only is allowed children or data not both
    children(props, propName) {
      const { data } = props;
      if (!props[propName] && !data) {
        return new Error('Invalid props, `data` or `children` is required');
      }
      if (data && data.length !== 0 && !props.renderItem) {
        return new Error('Invalid props, `renderItem` is required');
      }
      return undefined;
    },
  };

  static defaultProps = {
    numColumns: 3,
    data: [],
    itemStyle: {},
    showSeparator: false,
    showAnimation: false,
    animationDuration: 500,
    separatorBorderWidth: 0,
    separatorBorderColor: colors.white,
    animationInitialBackgroundColor: colors.white,
  };

  componentWillMount() {
    const {
      separatorBorderWidth,
      separatorBorderColor,
      numColumns,
      animationInitialBackgroundColor,
      showSeparator,
      showAnimation,
    } = this.props;

    const stylesOptions = {
      numColumns,
    };

    if (showSeparator) {
      stylesOptions.separatorBorderWidth = separatorBorderWidth;
      stylesOptions.separatorBorderColor = separatorBorderColor;
    }
    if (showAnimation) {
      stylesOptions.animationInitialBackgroundColor = animationInitialBackgroundColor;
    }
    this.styles = generateStyles(stylesOptions);

    this.setup(this.props);
    this.animate();
  }
  componentDidUpdate(nextProps) {
    this.setup(nextProps);
    this.animate();
  }

  setup = ({ children, data, renderItem }) => {
    if (children) {
      this._data = children;
      this._renderItem = this.renderChildren;
    } else if (data) {
      this._data = data;
      this._renderItem = renderItem;
    }
  };

  animate() {
    if (this.props.showAnimation) {
      const { numColumns, animationDuration } = this.props;
      this.animatedValue = [];

      this.animations = this._data.map((_, index) => {
        this.animatedValue[index] = new Animated.Value(0);
        return Animated.stagger(0, [
          Animated.timing(this.animatedValue[index], {
            toValue: 1,
            duration: animationDuration * Math.ceil((index + 1) / numColumns),
          }),
        ]);
      });
    }
  }

  _keyExtractor = (item, index) => index.toString();

  renderItem = ({ item, index }) => {
    const { showAnimation, showSeparator, itemStyle } = this.props;

    const viewStyles = [];
    viewStyles.push(this.styles.itemContainer);
    if (showSeparator) {
      viewStyles.push(this.styles.itemContainerSeparator);
    }
    if (showAnimation) {
      viewStyles.push(this.styles.itemContainerAnimationStart);
    }

    viewStyles.push(itemStyle);

    return (
      <View style={viewStyles}>
        {showAnimation ? (
          <Animated.View
            style={[
              this.styles.itemContainerAnimationEnd,
              { opacity: this.animatedValue[index] },
            ]}
          >
            {this._renderItem({
              item,
              index,
              animation: this.animations[index],
            })}
          </Animated.View>
        ) : (
          this._renderItem({ item, index })
        )}
      </View>
    );
  };

  renderChildren = ({ item, animation }) => {
    if (animation) {
      animation.start();
    }
    return item;
  };

  render() {
    const { showSeparator, ...props } = this.props;
    return (
      <FlatList
        contentContainerStyle={showSeparator && this.styles.container}
        keyExtractor={this._keyExtractor}
        ItemSeparatorComponent={() =>
          showSeparator ? <View style={this.styles.separator} /> : null
        }
        showsVerticalScrollIndicator={false}
        {...props}
        data={this._data}
        renderItem={this.renderItem}
      />
    );
  }
}
