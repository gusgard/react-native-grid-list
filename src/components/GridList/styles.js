import { width } from '../../themes';

export default ({
  numColumns,
  separatorBorderWidth,
  separatorBorderColor,
  animationInitialBackgroundColor,
}) => ({
  itemContainer: {
    maxHeight: width / numColumns,
    flex: 1,
  },
  itemContainerSeparator: {
    flex: 1,
    borderLeftWidth: separatorBorderWidth,
    borderColor: separatorBorderColor,
  },
  itemContainerAnimationStart: {
    backgroundColor: animationInitialBackgroundColor,
  },
  itemContainerAnimationEnd: {
    backgroundColor: separatorBorderColor,
    width: '100%',
    height: '100%',
  },
  separator: {
    borderColor: separatorBorderColor,
    borderWidth: separatorBorderWidth * 0.5,
  },
  container: {
    borderRightWidth: separatorBorderWidth,
    borderTopWidth: separatorBorderWidth,
    borderBottomWidth: separatorBorderWidth,
    borderColor: separatorBorderColor,
  },
});
