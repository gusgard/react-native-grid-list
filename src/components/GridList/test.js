import { shallow } from 'enzyme';
import React from 'react';
import { Image } from 'react-native';

import GridList from './index';

const logo = { uri: 'https://...' };
const items = [{ id: 1, thumbnail: logo }, { id: 2, thumbnail: logo }];

describe('grid list', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <GridList
        data={items}
        renderItem={({ item }) => <Image source={item.thumbnail} />}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
  
  it('renders empty list', () => {
    const wrapper = shallow(
      <GridList
        data={[]}
        renderItem={({ item }) => <Image source={item.thumbnail} />}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders showSeparator', () => {
    const wrapper = shallow(
      <GridList
        showSeparator
        data={items}
        renderItem={({ item }) => <Image source={item.thumbnail} />}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders showSeparator with border and width setted', () => {
    const wrapper = shallow(
      <GridList
        separatorBorderWidth={5}
        separatorBorderColor={'tomato'}
        showSeparator
        data={items}
        renderItem={({ item }) => <Image source={item.thumbnail} />}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders showAnimation', () => {
    const wrapper = shallow(
      <GridList
        showAnimation
        data={items}
        renderItem={({ item, stagger }) => (
          <Image source={item.thumbnail} onLoad={() => stagger.start()} />
        )}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders showAnimation with duration and initialBackground setted', () => {
    const wrapper = shallow(
      <GridList
        showAnimation
        animationInitialBackgroundColor={'beige'}
        animationDuration={1000}
        data={items}
        renderItem={({ item, stagger }) => (
          <Image source={item.thumbnail} onLoad={() => stagger.start()} />
        )}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders children', () => {
    const wrapper = shallow(
      <GridList>
        <Image source={items[0].thumbnail} />
        <Image source={items[1].thumbnail} />
      </GridList>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
