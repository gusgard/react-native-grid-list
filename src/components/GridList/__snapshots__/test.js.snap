// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grid list renders children 1`] = `
<FlatList
  ItemSeparatorComponent={[Function]}
  animationDuration={500}
  animationInitialBackgroundColor="#FFFFFF"
  contentContainerStyle={false}
  data={
    Array [
      <Image
        source={
          Object {
            "uri": "https://...",
          }
        }
      />,
      <Image
        source={
          Object {
            "uri": "https://...",
          }
        }
      />,
    ]
  }
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  itemStyle={Object {}}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={3}
  onEndReachedThreshold={2}
  renderItem={[Function]}
  scrollEventThrottle={50}
  separatorBorderColor="#FFFFFF"
  separatorBorderWidth={0}
  showAnimation={false}
  showsVerticalScrollIndicator={false}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <Image
    source={
      Object {
        "uri": "https://...",
      }
    }
  />
  <Image
    source={
      Object {
        "uri": "https://...",
      }
    }
  />
</FlatList>
`;

exports[`grid list renders correctly 1`] = `
<FlatList
  ItemSeparatorComponent={[Function]}
  animationDuration={500}
  animationInitialBackgroundColor="#FFFFFF"
  contentContainerStyle={false}
  data={
    Array [
      Object {
        "id": 1,
        "thumbnail": Object {
          "uri": "https://...",
        },
      },
      Object {
        "id": 2,
        "thumbnail": Object {
          "uri": "https://...",
        },
      },
    ]
  }
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  itemStyle={Object {}}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={3}
  onEndReachedThreshold={2}
  renderItem={[Function]}
  scrollEventThrottle={50}
  separatorBorderColor="#FFFFFF"
  separatorBorderWidth={0}
  showAnimation={false}
  showsVerticalScrollIndicator={false}
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;

exports[`grid list renders empty list 1`] = `
<FlatList
  ItemSeparatorComponent={[Function]}
  animationDuration={500}
  animationInitialBackgroundColor="#FFFFFF"
  contentContainerStyle={false}
  data={Array []}
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  itemStyle={Object {}}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={3}
  onEndReachedThreshold={2}
  renderItem={[Function]}
  scrollEventThrottle={50}
  separatorBorderColor="#FFFFFF"
  separatorBorderWidth={0}
  showAnimation={false}
  showsVerticalScrollIndicator={false}
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;

exports[`grid list renders showAnimation 1`] = `
<FlatList
  ItemSeparatorComponent={[Function]}
  animationDuration={500}
  animationInitialBackgroundColor="#FFFFFF"
  contentContainerStyle={false}
  data={
    Array [
      Object {
        "id": 1,
        "thumbnail": Object {
          "uri": "https://...",
        },
      },
      Object {
        "id": 2,
        "thumbnail": Object {
          "uri": "https://...",
        },
      },
    ]
  }
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  itemStyle={Object {}}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={3}
  onEndReachedThreshold={2}
  renderItem={[Function]}
  scrollEventThrottle={50}
  separatorBorderColor="#FFFFFF"
  separatorBorderWidth={0}
  showAnimation={true}
  showsVerticalScrollIndicator={false}
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;

exports[`grid list renders showAnimation with duration and initialBackground setted 1`] = `
<FlatList
  ItemSeparatorComponent={[Function]}
  animationDuration={1000}
  animationInitialBackgroundColor="beige"
  contentContainerStyle={false}
  data={
    Array [
      Object {
        "id": 1,
        "thumbnail": Object {
          "uri": "https://...",
        },
      },
      Object {
        "id": 2,
        "thumbnail": Object {
          "uri": "https://...",
        },
      },
    ]
  }
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  itemStyle={Object {}}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={3}
  onEndReachedThreshold={2}
  renderItem={[Function]}
  scrollEventThrottle={50}
  separatorBorderColor="#FFFFFF"
  separatorBorderWidth={0}
  showAnimation={true}
  showsVerticalScrollIndicator={false}
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;

exports[`grid list renders showSeparator 1`] = `
<FlatList
  ItemSeparatorComponent={[Function]}
  animationDuration={500}
  animationInitialBackgroundColor="#FFFFFF"
  contentContainerStyle={
    Object {
      "borderBottomWidth": 0,
      "borderColor": "#FFFFFF",
      "borderRightWidth": 0,
      "borderTopWidth": 0,
    }
  }
  data={
    Array [
      Object {
        "id": 1,
        "thumbnail": Object {
          "uri": "https://...",
        },
      },
      Object {
        "id": 2,
        "thumbnail": Object {
          "uri": "https://...",
        },
      },
    ]
  }
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  itemStyle={Object {}}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={3}
  onEndReachedThreshold={2}
  renderItem={[Function]}
  scrollEventThrottle={50}
  separatorBorderColor="#FFFFFF"
  separatorBorderWidth={0}
  showAnimation={false}
  showsVerticalScrollIndicator={false}
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;

exports[`grid list renders showSeparator with border and width setted 1`] = `
<FlatList
  ItemSeparatorComponent={[Function]}
  animationDuration={500}
  animationInitialBackgroundColor="#FFFFFF"
  contentContainerStyle={
    Object {
      "borderBottomWidth": 5,
      "borderColor": "tomato",
      "borderRightWidth": 5,
      "borderTopWidth": 5,
    }
  }
  data={
    Array [
      Object {
        "id": 1,
        "thumbnail": Object {
          "uri": "https://...",
        },
      },
      Object {
        "id": 2,
        "thumbnail": Object {
          "uri": "https://...",
        },
      },
    ]
  }
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  itemStyle={Object {}}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={3}
  onEndReachedThreshold={2}
  renderItem={[Function]}
  scrollEventThrottle={50}
  separatorBorderColor="tomato"
  separatorBorderWidth={5}
  showAnimation={false}
  showsVerticalScrollIndicator={false}
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;
